### PR TITLE
Feature/teamwork

### PR DESCRIPTION
**Pull request by luiskcs89**

### CODE REVIEW

File: functions/src/services/teamwork.service.ts [hunk 1]
```diff
@@ -28,6 +28,43 @@ export class TeamworkService {
       .catch(error => { throw new Error('Unable to authenticate with Teamwork. Please verify your token and try again later.') });
   }
 
+  async getProjects(token: string) {
+    let authData;
+    try {
+      authData = await this.validateToken(token);
```
Comment (kshaaban-) [inline functions/src/services/teamwork.service.ts]: We should not need to make a request to the authenticate endpoint to obtain `authData`.

The endpoint GET `connections/:connectionId/projects` will have the `uid` (from authentication middleware) and the `connectionId` from the route params.

Using these two, we can use the firestore service to get the connection details from document at this path `/users/${uid}/connections/${connectionId}`, which will provide the `token` and `authData` used in this method.

File: functions/src/services/teamwork.service.ts [hunk 2]
```diff
@@ -65,6 +65,46 @@ export class TeamworkService {
     }
   }
 
+  async getTaskLists(token: string, projectId: string, page: number) {
+    let authData;
+    try {
+      authData = await this.validateToken(token);
```
Comment (kshaaban-) [inline functions/src/services/teamwork.service.ts]: Similar to the previous comment, a GET on `connections/:connectionId/projects/:projectId/task-lists`
should have enough data in the request handler to provide token/projectId to this method.

File: functions/src/services/teamwork.service.ts [hunk 3]
```diff
@@ -105,6 +105,86 @@ export class TeamworkService {
...
+        t.childTasks = allTasks.filter(c => c.parent == t.id);
+        if(t.childTasks.length > 0) {
+          this.fillChildTasks(t.childTasks, allTasks);
+        } 
+      }
+    );
+  }
+
+  private async getTasksData(token: string, tasklistId: string, page: number) {
+    let authData;
```
Comment (kshaaban-) [inline functions/src/services/teamwork.service.ts]: See previous two comments

**Comment by luiskcs89** [comment]
Added projects, task lists and tasks endpoints that use teamwork service and fixed team work service to use connection data.

**Comment by kshaaban-** [comment]
Last two small but very important things:

### Resource Naming
Since we've named all resources as plurals, the `task` endpoints should be renamed to `tasks`

### Response Format
When returning `projects`, the response is `{ projects: Project[] }`. But when returning tasks, the response is `Task[]`. I believe it's better to return an object because it can be extended later on (if we need to send meta-data like pagination or whatever else may change) without having to make changes on the consuming code on the client

**Pull request merged**